border0/resource_socket.go

```diff
@@ -0,0 +1,382 @@
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the socket.",
+			},
+			"socket_type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The type of the socket.",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The description of the socket.",
+			},
+			"recording_enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Indicates if recording is enabled for the socket.",
+			},
+			"connector_authentication_enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Indicates if connector authentication is enabled for the socket.",
+			},
+			"tags": {
+				Type: schema.TypeMap,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional:    true,
+				Description: "The tags of the socket.",
+			},
+			"connector_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The connector id that the socket is associated with.",
+			},
+			"upstream_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					if old != "" && new == "" {
+						return true
+					}
+					return old == new
+				},
+				Description: "The upstream type of the socket.",
+			},
+			"upstream_http_hostname": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The upstream http hostname of the socket.",
+			},
+			"upstream_connection_type": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The upstream connection type. Valid values: `ssh`, `aws_ec2_connect`, `aws_ssm`, `database`. Defaults to `ssh`.",
+				Default:     border0types.UpstreamConnectionTypeSSH,
+			},
+			"upstream_hostname": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The upstream hostname.",
+			},
+			"upstream_port": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "The upstream port number.",
+			},
+			"upstream_authentication_type": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The upstream authentication type. Valid values: `username_password`, `border0_cert`, `ssh_private_key`. Defaults to `border0_cert`.",
+				Default:     border0types.UpstreamAuthenticationTypeBorder0Cert,
+			},
+			"upstream_username": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Sensitive:   true,
+				Description: "The upstream username.",
+			},
+			"upstream_password": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Sensitive:   true,
+				Description: "The upstream password.",
+			},
+			"upstream_private_key": {
```

Comment:
Still not a fan of this flat structure...

Reply:
Yeah, I think we should have some nested schema. I have started the work, but I removed the code to keep this PR smaller.